core/drivers/firewall/stm32_risaf.c

```diff
@@ -0,0 +1,770 @@
+};
+
+struct stm32_risaf_pdata {
+	struct io_pa_va base;
+	struct clk *clock;
+	struct stm32_risaf_region *regions;
+	char risaf_name[20];
+	unsigned int nregions;
+	unsigned int conf_lock;
+	uintptr_t mem_base;
+	size_t mem_size;
+	bool enc_supported;
+};
+
+struct stm32_risaf_ddata {
+	uint32_t mask_regions;
+	uint32_t max_base_regions;
+	uint32_t granularity;
+};
+
+struct stm32_risaf_instance {
+	struct stm32_risaf_pdata pdata;
+	struct stm32_risaf_ddata *ddata;
+
+	SLIST_ENTRY(stm32_risaf_instance) link;
+};
+
+struct stm32_risaf_version {
+	uint32_t major;
+	uint32_t minor;
+	uint32_t ip_id;
+	uint32_t size_id;
+};
+
+/**
+ * struct stm32_risaf_compat_data describes RISAF associated data
+ * for compatible list.
+ *
+ * @supported_encryption:	identify risaf encryption capabilities.
+ */
+struct stm32_risaf_compat_data {
+	bool supported_encryption;
+};
+
+static bool is_tdcid;
+
+static const struct stm32_risaf_compat_data stm32_risaf_compat = {
+	.supported_encryption = false,
+};
+
+static const struct stm32_risaf_compat_data stm32_risaf_enc_compat = {
+	.supported_encryption = true,
+};
+
+static SLIST_HEAD(, stm32_risaf_instance) risaf_list =
+		SLIST_HEAD_INITIALIZER(risaf_list);
+
+static vaddr_t risaf_base(struct stm32_risaf_instance *risaf)
+{
+	return io_pa_or_va_secure(&risaf->pdata.base, 1);
+}
+
+void stm32_risaf_clear_illegal_access_flags(void)
+{
+	struct stm32_risaf_instance *risaf = NULL;
+
+	SLIST_FOREACH(risaf, &risaf_list, link) {
+		vaddr_t base = io_pa_or_va_secure(&risaf->pdata.base, 1);
+
+		if (clk_enable(risaf->pdata.clock))
+			panic("Can't enable RISAF clock");
+
+		if (!io_read32(base + _RISAF_IASR)) {
+			clk_disable(risaf->pdata.clock);
+			continue;
+		}
+
+		io_write32(base + _RISAF_IACR, _RISAF_IACR_CAEF |
+			   _RISAF_IACR_IAEF0 | _RISAF_IACR_IAEF1);
+
+		clk_disable(risaf->pdata.clock);
+	}
+}
+
+#ifdef CFG_TEE_CORE_DEBUG
+void stm32_risaf_dump_erroneous_data(void)
+{
+	struct stm32_risaf_instance *risaf = NULL;
+
+	SLIST_FOREACH(risaf, &risaf_list, link) {
+		vaddr_t base = io_pa_or_va_secure(&risaf->pdata.base, 1);
+
+		if (clk_enable(risaf->pdata.clock))
+			panic("Can't enable RISAF clock");
+
+		/* Check if faulty address on this RISAF */
+		if (!io_read32(base + _RISAF_IASR)) {
+			clk_disable(risaf->pdata.clock);
+			continue;
+		}
+
+		EMSG("\n\nDUMPING DATA FOR %s\n\n", risaf->pdata.risaf_name);
+		EMSG("=====================================================");
+		EMSG("Status register (IAESR0): %#"PRIx32,
+		     io_read32(base + _RISAF_IAESR0));
+
+		/* Reserved if dual port feature not available */
+		if (io_read32(base + _RISAF_IAESR1))
+			EMSG("Status register Dual Port (IAESR1) %#"PRIx32,
+			     io_read32(base + _RISAF_IAESR1));
+
+		EMSG("-----------------------------------------------------");
+		if (virt_to_phys((void *)base) == RISAF4_BASE) {
+			EMSG("Faulty address (IADDR0): %#"PRIxPA,
+			     risaf->pdata.mem_base +
+			     (paddr_t)io_read32(base + _RISAF_IADDR0));
```

Comment:
Cast to `(paddr_t)` not needed.

This seems a real error case, I think the function should panicbefore exiting if 1 faulty config is found.